include/observable.h

```diff
@@ -0,0 +1,271 @@
+ * Purpose: General observable implementation with several specializations.
+ *
+ * Copyright (C) 2022 Alec Leamas
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,  USA.
+ **************************************************************************/
+
+#ifndef OBSERVABLE_H
+#define OBSERVABLE_H
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <wx/config.h>
+#include <wx/event.h>
+
+/** Return address as printable string. */
+std::string ptr_key(const void* ptr);
+
+
+class ObservedVar;
+class ObservedVarListener;
+
+
+/**
+ *  Private helper class. Basically a singleton map of listeners where
+ *  singletons are managed by key, one for each key value.
+ */
+class SingletonVar {
+
+private:
+  static SingletonVar* getInstance(const std::string& key);
+  std::map<wxEvtHandler*, wxEventType> listeners;
+
+  SingletonVar() {}
+  SingletonVar(const SingletonVar&);    // not implemented
+  void operator=(const SingletonVar&);  // not implemented
+
+  friend class ObservedVar;
+};
+
+
+/**  The observable notify/listen basic nuts and bolts.  */
+class ObservedVar {
+public:
+  ObservedVar(const std::string& key)
+      : singleton(SingletonVar::getInstance(key)) {}
+
+  /** Notify all listeners about variable change. */
+  virtual const void notify();
+
+  /** Remove window from list of listeners, return true if listener exists */
+  bool unlisten(wxEvtHandler* listener);
+
+
+  /** Get listening object which unlisten() on destruction. */
+  typedef std::unique_ptr<ObservedVarListener> Listener;
+  Listener get_listener(wxEvtHandler* handler, wxEventType ev);
+
+
+protected:
+  /**
+   * Notify all listeners: send them a 'type' wxCommandEvent message
+   * as defined by listen() with optional data available using GetString()
+   * and/or GetClientData().
+   */
+  const void notify(const std::string& s, void* client_data);
+
+private:
+  /** Set object to send ev_type to listener on variable changes. */
+  void listen(wxEvtHandler* listener, wxEventType ev_type);
+
+  SingletonVar* const singleton;
+
+  friend class ObservedVarListener;
+};
+
+
+/**
+ *  Keeps listening over it's lifespan, removes itself on destruction.
+ *  Basically private, only available through ObservedVar::get_listener().
+ */
+class ObservedVarListener {
```

Comment:
Mark the class `final`

Reply:
Also, I would make this a member class that can be just called `Listener` inside `ObservedVar`.

Reply:
final is missing, agreed (fixed).

Making it an inner class is complicated since it contains a reference to the ObservedVad outer class.  There are all sorts of work-arounds, but having it as a separate class is perhaps the simplest way to deal with it.